### PR TITLE
Convert (remaining non-Peer-related) to CompletableFuture

### PR DESCRIPTION
**Pull request by msgilligan**
This converts all/most of the remaining non Peer-related code in `core` to use `CompletableFuture`.

### CODE REVIEW

File: core/src/test/java/org/bitcoinj/core/ECKeyTest.java [hunk 1]
```diff
@@ -72,15 +70,18 @@ public void setUp() throws Exception {
     public void sValue() throws Exception {
         // Check that we never generate an S value that is larger than half the curve order. This avoids a malleability
         // issue that can allow someone to change a transaction [hash] without invalidating the signature.
-        final int ITERATIONS = 10;
```
Comment (schildbach) [inline core/src/test/java/org/bitcoinj/core/ECKeyTest.java:75]: Why this change to `byte`? Feels weird to keep a count in a byte.

File: core/src/test/java/org/bitcoinj/core/ECKeyTest.java [hunk 2]
```diff
@@ -72,15 +70,18 @@ public void setUp() throws Exception {
...
-            sigFutures.add(executor.submit(() -> key.sign(hash)));
-        }
-        List<ECKey.ECDSASignature> sigs = Futures.allAsList(sigFutures).get();
+
+        Function<Byte, ECKey.ECDSASignature> signer = b -> key.sign(Sha256Hash.of(new byte[]{b}));
+        Function<Byte, CompletableFuture<ECKey.ECDSASignature>> asyncSigner = b -> CompletableFuture.supplyAsync(() -> signer.apply(b));
+
+        List<CompletableFuture<ECKey.ECDSASignature>> sigFutures = IntStream.range(0, ITERATIONS - 1)
+                .mapToObj(i -> (byte) i)
+                .map(asyncSigner)
```
Comment (schildbach) [inline core/src/test/java/org/bitcoinj/core/ECKeyTest.java:81]: Seems to me that instead if signing `hash(i)` we're now signing just `i`. Is that intentional? It probably doesn't matter much though.

File: core/src/test/java/org/bitcoinj/core/ECKeyTest.java [hunk 3]
```diff
@@ -72,15 +70,18 @@ public void setUp() throws Exception {
     public void sValue() throws Exception {
         // Check that we never generate an S value that is larger than half the curve order. This avoids a malleability
```
Comment (schildbach) [inline core/src/test/java/org/bitcoinj/core/ECKeyTest.java:71]: For both the old and the new code I miss where the actual "check" is executed. Do you know?

**Review by schildbach — COMMENTED** [review]
I merged most of your changes as separate commits, or will merge soon. For ECKeyTest, I've added a few review comments.

File: core/src/test/java/org/bitcoinj/core/ECKeyTest.java (hunk 1 above)
Comment (msgilligan) [reply core/src/test/java/org/bitcoinj/core/ECKeyTest.java:75]: Because the code that generates the hashes uses a single-byte buffer and this value shouldn't be made larger than a byte without some consideration.

File: core/src/test/java/org/bitcoinj/core/ECKeyTest.java (hunk 2 above)
Comment (msgilligan) [reply core/src/test/java/org/bitcoinj/core/ECKeyTest.java:81]: The `.map(asyncSigner)` step calls the `asyncSinger` function, which calls the `signer` function which does the hash.

File: core/src/test/java/org/bitcoinj/core/ECKeyTest.java (hunk 3 above)
Comment (msgilligan) [reply core/src/test/java/org/bitcoinj/core/ECKeyTest.java:71]: The checks are performed _after_ the async calculations complete and the results of the completed futures are assembled in a list. This starts at line 85 in the "after" code of this PR:

```
for (ECKey.ECDSASignature signature : sigs) {
     assertTrue(signature.isCanonical());
}
```

File: core/src/test/java/org/bitcoinj/core/ECKeyTest.java (hunk 2 above)
Comment (schildbach) [reply core/src/test/java/org/bitcoinj/core/ECKeyTest.java:81]: Ah, I missed the Sha256Hash in the function.

File: core/src/test/java/org/bitcoinj/core/ECKeyTest.java [hunk 4]
```diff
@@ -72,15 +70,18 @@ public void setUp() throws Exception {
...
-        for (byte i = 0; i < ITERATIONS; i++) {
-            final Sha256Hash hash = Sha256Hash.of(new byte[]{i});
-            sigFutures.add(executor.submit(() -> key.sign(hash)));
-        }
-        List<ECKey.ECDSASignature> sigs = Futures.allAsList(sigFutures).get();
+
+        Function<Byte, ECKey.ECDSASignature> signer = b -> key.sign(Sha256Hash.of(new byte[]{b}));
+        Function<Byte, CompletableFuture<ECKey.ECDSASignature>> asyncSigner = b -> CompletableFuture.supplyAsync(() -> signer.apply(b));
+
+        List<CompletableFuture<ECKey.ECDSASignature>> sigFutures = IntStream.range(0, ITERATIONS - 1)
```
Comment (schildbach) [inline core/src/test/java/org/bitcoinj/core/ECKeyTest.java:79]: I think no `- 1` needed here. The end is exclusive, just as with the loop before.

File: core/src/test/java/org/bitcoinj/core/ECKeyTest.java (hunk 1 above)
Comment (schildbach) [reply core/src/test/java/org/bitcoinj/core/ECKeyTest.java:75]: Ok, I see.

Ultimately as I understand the code it just feeds pseudo-random data into the signer.

But for the purpose of this refactoring, it's fine I guess.

File: core/src/test/java/org/bitcoinj/core/ECKeyTest.java (hunk 3 above)
Comment (schildbach) [reply core/src/test/java/org/bitcoinj/core/ECKeyTest.java:71]: You're right, thanks for pointing this out!

**Comment by schildbach** [comment]
Merged.
